### PR TITLE
Cisco telnet - error handling during login

### PR DESCRIPTION
**Pull request by ddisisto**
This is related to #621 - I found that the reason it was retrying the same (invalid) credentials was because the output returned by read_channel() looked like this:
```
% Login invalid

Username:
Username:
```

Which was identified as just another Username: prompt hence the username was sent, until `max_loops` is eventually reached.

The Cisco standard for any problems during telnet login is to output:
```
% Error message
```
Where the actual error message would be something like: `Password required, but none set`, `Authentication failed`, `Login invalid`, etc.

This PR makes the error handling more generic by looking for for that type of error output at all stages during the login process, and raising `NetMikoAuthenticationException` containing the actual error message if any encountered.

I've also removed the one special case of error handing that was previously done, as this is now handled in effectively the same way by the new method.

### CODE REVIEW

**Comment by zeepi222** [comment]
Thank you @ddisisto this has resolved the issue I was experiencing.

**Comment by ktbyers** [comment]
@ddisisto Yes, I intentionally want to try to login via telnet multiple times. 

Basically, telnet login is not sufficiently reliable to do this in a single attempt. It needs to try multiple times. This is especially true since this also needs to handle telnet through terminal servers and serial port-console connections. 

Even without those cases I am pretty leery that a single telnet login attempt would be sufficiently reliable.

Basically, my main focus is to reliably login.


What is the negative for the failed logins? @ddisisto mentioned in #621 potentially having the account locked in TACACS+, is that your main concern here?

**Comment by ktbyers** [comment]
I am going to close this. I am not seeing what problem we are trying to solve (I assuming locking out TACACS+ account)? The multiple login attempts is by design (to make the login more reliable).

**Comment by ktbyers** [comment]
@zeepi222 No, that is incorrect. Netmiko will only have a single telnet connection (a single TCP socket). If that telnet connection dies, Netmiko will fail to connect (but it will only take up a single VTY port).

**Comment by zeepi222** [comment]
@ktbyers 
Yes, understood. Sorry I didn't explain the situation clearly. The issue is not the connection dying, the initial telnet connection is established fine. But when netmiko is unable to enter enable mode (due to an invalid secret),  netmiko raises a Value error and doesn't end the cli_session cleanly, this leaves the vty connection in a used state. The scenario I described previously occurred after running the netmiko script 5 times.

Here is an example showing the before and after conditions, when using an invalid secret password.

**Before** (note that only one vty session is used, which is the current telnet connection - line 388):
```
/etc/ansible # telnet 10.179.183.1
Entering character mode
Escape character is '^]'.

User Access Verification

Username: mwadmin
Password: 

kanrou000001>enable
Password: 
kanrou000001#who
    Line       User       Host(s)              Idle       Location
*388 vty 1     mwadmin   idle                 00:00:00 10.179.255.250

  Interface    User               Mode         Idle     Peer Address

kanrou000001#exit
Connection closed by foreign host
/etc/ansible # 
```

**After** (after running the netmiko script you will notice two vty sessions are used, the netmiko session which didn't disconnect cleanly line 388 and the current telnet connection - line 389):

```
/etc/ansible # python telnet_test.py 
25-Feb-2018 00:47:39 read_channel: C

User Access Verification

Username: 
25-Feb-2018 00:47:39 write_channel: mwadmin

25-Feb-2018 00:47:43 read_channel: mwadmin
Password: 
25-Feb-2018 00:47:43 write_channel: jump3r5t@rt

25-Feb-2018 00:47:46 read_channel: 

kanrou000001>
25-Feb-2018 00:47:50 read_channel: 
25-Feb-2018 00:47:58 read_channel: 
25-Feb-2018 00:47:58 write_channel: 

25-Feb-2018 00:47:59 read_channel: 
25-Feb-2018 00:48:07 read_channel: 
kanrou000001>
25-Feb-2018 00:48:07 read_channel: 
25-Feb-2018 00:48:15 read_channel: 
25-Feb-2018 00:48:15 read_channel: 
25-Feb-2018 00:48:15 write_channel: 

25-Feb-2018 00:48:16 read_channel: 
25-Feb-2018 00:48:16 read_channel: 
25-Feb-2018 00:48:16 write_channel: 

25-Feb-2018 00:48:16 read_channel: 
kanrou000001>
25-Feb-2018 00:48:16 read_channel: 
25-Feb-2018 00:48:17 read_channel: 
kanrou000001>
25-Feb-2018 00:48:17 In disable_paging
25-Feb-2018 00:48:17 Command: terminal length 0

25-Feb-2018 00:48:17 write_channel: terminal length 0

25-Feb-2018 00:48:17 Pattern is: kanrou000001
25-Feb-2018 00:48:17 read_channel: 
25-Feb-2018 00:48:17 read_channel: 
25-Feb-2018 00:48:18 read_channel: terminal length 0
kanrou000001>
25-Feb-2018 00:48:18 Pattern found: kanrou000001 terminal length 0
kanrou000001>
25-Feb-2018 00:48:18 terminal length 0
kanrou000001>
25-Feb-2018 00:48:18 Exiting disable_paging
25-Feb-2018 00:48:18 write_channel: terminal width 511

25-Feb-2018 00:48:18 Pattern is: kanrou000001
25-Feb-2018 00:48:18 read_channel: 
25-Feb-2018 00:48:18 read_channel: 
25-Feb-2018 00:48:18 read_channel: terminal 
25-Feb-2018 00:48:19 read_channel: 
25-Feb-2018 00:48:19 read_channel: 
25-Feb-2018 00:48:20 read_channel: width 511
kanrou000001>
25-Feb-2018 00:48:20 Pattern found: kanrou000001 terminal width 511
kanrou000001>
25-Feb-2018 00:48:21 read_channel: 
25-Feb-2018 00:48:21 write_channel: 

25-Feb-2018 00:48:21 Pattern is: kanrou000001
25-Feb-2018 00:48:21 read_channel: 
25-Feb-2018 00:48:21 read_channel: 
25-Feb-2018 00:48:22 read_channel: 
kanrou000001>
25-Feb-2018 00:48:22 Pattern found: kanrou000001 
kanrou000001>
25-Feb-2018 00:48:22 write_channel: enable

25-Feb-2018 00:48:22 Pattern is: (kanrou000001|ssword)
25-Feb-2018 00:48:22 read_channel: 
25-Feb-2018 00:48:22 read_channel: 
25-Feb-2018 00:48:22 read_channel: enable
Password: 
25-Feb-2018 00:48:22 Pattern found: (kanrou000001|ssword) enable
Password: 
25-Feb-2018 00:48:22 write_channel: invalid

25-Feb-2018 00:48:22 Pattern is: kanrou000001
25-Feb-2018 00:48:22 read_channel: 
25-Feb-2018 00:48:23 read_channel: 
25-Feb-2018 00:48:23 read_channel: 
% Access denied

kanrou000001>
25-Feb-2018 00:48:23 Pattern found: kanrou000001 
% Access denied

kanrou000001>
25-Feb-2018 00:48:23 write_channel: 

25-Feb-2018 00:48:23 Pattern is: kanrou000001
25-Feb-2018 00:48:23 read_channel: 
25-Feb-2018 00:48:24 read_channel: 
25-Feb-2018 00:48:24 read_channel: 
kanrou000001>
25-Feb-2018 00:48:24 Pattern found: kanrou000001 
kanrou000001>
Traceback (most recent call last):
  File "telnet_test.py", line 22, in <module>
    cli_session.enable()
  File "/usr/lib/python2.7/site-packages/netmiko/cisco_base_connection.py", line 18, in enable
    return super(CiscoBaseConnection, self).enable(cmd=cmd, pattern=pattern, re_flags=re_flags)
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 1049, in enable
    raise ValueError(msg)
ValueError: Failed to enter enable mode. Please ensure you pass the 'secret' argument to ConnectHandler.
/etc/ansible # 
/etc/ansible # telnet 10.179.183.1

Entering character mode
Escape character is '^]'.

User Access Verification

Username: mwadmin
Password: 

kanrou000001>enable
Password: 
kanrou000001#who
    Line       User       Host(s)              Idle       Location
 388 vty 0     mwadmin   idle                 00:00:30 10.179.255.250
*389 vty 1     mwadmin   idle                 00:00:00 10.179.255.250

  Interface    User               Mode         Idle     Peer Address

kanrou000001#exit
Connection closed by foreign host
/etc/ansible # 

```

Should I log this as a separate bug?

**Comment by ktbyers** [comment]
Are you calling the netmiko .disconnect() method in your script?

**Comment by zeepi222** [comment]
Here is the test script I'm using ...

```
import logging
from netmiko import ConnectHandler
logging.basicConfig(format='%(asctime)s %(message)s', datefmt='%d-%b-%Y %H:%M:%S', level=logging.DEBUG)
logging.getLogger("paramiko.transport").setLevel(logging.DEBUG)
logging.getLogger("netmiko").setLevel(logging.DEBUG)

ip = "10.179.183.1"
username = "mwadmin"
password = "jump3r5t@rt"
secret = "invalid"

cli_session = ConnectHandler(ip=ip,
                             username=username,
                             password=password,
                             secret=secret,
                             device_type='cisco_ios_telnet',
                             global_delay_factor=4)

cli_session.enable()
device_prompt = cli_session.find_prompt()
print('My Prompt is: {}'.format(device_prompt))
cmd = cli_session.send_command('show version')
cli_session.send_command('exit')
cli_session.disconnect()
print(cmd)

```

Because the script never gets into enable mode, it never gets past the line that print's the prompt.

**Comment by ktbyers** [comment]
Your `secret` has to be right...otherwise, you will get an exception on the `enable()` method. 

Since you are getting an exception, your script is not calling the `disconnect()` method (consequently, you are not gracefully closing the SSH session).

So either use a context manager, exception handling, or make sure your secret is correct.

**Comment by zeepi222** [comment]
Unfortunately our automation pulls in credentials from an external source, so the secret is not always guaranteed to be correct.

I tried handling the ValueError exception but that results in a EOFError exception.

Can you recommend a simple example of how to address this?

This is the how I configured the exception handling:

```
import logging
from netmiko import ConnectHandler
from netmiko.ssh_exception import NetMikoTimeoutException, NetMikoAuthenticationException, SSHException
logging.basicConfig(format='%(asctime)s %(message)s', datefmt='%d-%b-%Y %H:%M:%S', level=logging.DEBUG)
logging.getLogger("paramiko.transport").setLevel(logging.DEBUG)
logging.getLogger("netmiko").setLevel(logging.DEBUG)

ip = "10.179.183.1"
username = "mwadmin"
password = "jumper"
secret = "invalid"

try:
    cli_session = ConnectHandler(ip=ip,
                                 username=username,
                                 password=password,
                                 secret=secret,
                                 device_type='cisco_ios_telnet',
                                 global_delay_factor=4,
                                 timeout=120)

    cli_session.enable()
    cli_session.send_command('show version')
    cli_session.disconnect()

except ValueError as err:
    print('*** Unable to enter enable mode ***'.format(err))
    cli_session.send_command('exit')
    cli_session.disconnect()

``` 

The output looks like this:

```
kanrou000001>
*** Unable to enter enable mode ***
25-Feb-2018 02:31:24 read_channel: 
25-Feb-2018 02:31:24 write_channel: 
25-Feb-2018 02:31:24 read_channel: 
25-Feb-2018 02:31:24 read_channel: 
25-Feb-2018 02:31:24 write_channel: 
25-Feb-2018 02:31:25 read_channel: 
kanrou000001>
25-Feb-2018 02:31:25 read_channel: 
kanrou000001>
25-Feb-2018 02:31:26 read_channel: 
25-Feb-2018 02:31:26 write_channel: exit

25-Feb-2018 02:31:26 read_channel: 
25-Feb-2018 02:31:27 read_channel: ex
25-Feb-2018 02:31:27 read_channel: 
25-Feb-2018 02:31:27 read_channel: 
25-Feb-2018 02:31:28 read_channel: it

Traceback (most recent call last):
  File "telnet_test.py", line 28, in <module>
    cli_session.send_command('exit')
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 945, in send_command
    new_data = self.read_channel()
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 329, in read_channel
    output = self._read_channel()
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 316, in _read_channel
    output = self.remote_conn.read_very_eager().decode('utf-8', 'ignore')
  File "/usr/lib/python2.7/telnetlib.py", line 418, in read_very_eager
    return self.read_very_lazy()
  File "/usr/lib/python2.7/telnetlib.py", line 455, in read_very_lazy
    raise EOFError, 'telnet connection closed'
EOFError: telnet connection closed

```

**Comment by ktbyers** [comment]
Yes, get rid of this:

```
   cli_session.send_command('exit')
```

disconnect() handles this.

**Comment by zeepi222** [comment]
@ktbyers thank  you. Removing that line did in one instance help achieve a cleaner disconnection of the session. Unfortunately, I can't seem to get consistent results. Out of **3** attempts, only **1** attempt responded as expected. For clients like this one which have slower WAN connections, I have tried bumping up the connection timers to try and compensate for increased latency. When connecting manually, the cli response is slow but always returns the expected response.

**Code**
```

import logging
from netmiko import ConnectHandler
from netmiko.ssh_exception import NetMikoTimeoutException, NetMikoAuthenticationException, SSHException
logging.basicConfig(format='%(asctime)s %(message)s', datefmt='%d-%b-%Y %H:%M:%S', level=logging.DEBUG)
logging.getLogger("paramiko.transport").setLevel(logging.DEBUG)
logging.getLogger("netmiko").setLevel(logging.DEBUG)

ip = "10.179.183.1"
username = "mwadmin"
password = "jump3r5t@rt"
secret = "invalid"

try:
    cli_session = ConnectHandler(ip=ip,
                                 username=username,
                                 password=password,
                                 secret=secret,
                                 device_type='cisco_ios_telnet',
                                 global_delay_factor=4,
                                 timeout=90,
                                 blocking_timeout=90,
                                 session_timeout=90)

    cli_session.enable()
    cli_session.send_command('show version')
    cli_session.disconnect()

except ValueError as err:
    cli_session.disconnect()
    raise NetMikoAuthenticationException('{}'.format(err))


```
**First attempt** - failed within a few seconds
```

/etc/ansible # python telnet_test.py 
25-Feb-2018 06:50:18 read_channel: 
25-Feb-2018 06:50:18 write_channel: 

25-Feb-2018 06:50:20 read_channel: 
25-Feb-2018 06:50:20 write_channel: 

25-Feb-2018 06:50:22 read_channel: 
25-Feb-2018 06:50:22 write_channel: 

25-Feb-2018 06:50:24 read_channel: 
25-Feb-2018 06:50:24 write_channel: 

25-Feb-2018 06:50:26 read_channel: 
25-Feb-2018 06:50:26 write_channel: 

25-Feb-2018 06:50:28 read_channel: 
25-Feb-2018 06:50:28 write_channel: 

25-Feb-2018 06:50:30 read_channel: 
25-Feb-2018 06:50:30 write_channel: 

25-Feb-2018 06:50:32 read_channel: 
25-Feb-2018 06:50:32 write_channel: 

Traceback (most recent call last):
  File "telnet_test.py", line 22, in <module>
    session_timeout=90)
  File "/usr/lib/python2.7/site-packages/netmiko/ssh_dispatcher.py", line 173, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 160, in __init__
    self.establish_connection()
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 640, in establish_connection
    self.telnet_login()
  File "/usr/lib/python2.7/site-packages/netmiko/cisco_base_connection.py", line 78, in telnet_login
    output = self.read_channel()
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 329, in read_channel
    output = self._read_channel()
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 316, in _read_channel
    output = self.remote_conn.read_very_eager().decode('utf-8', 'ignore')
  File "/usr/lib/python2.7/telnetlib.py", line 416, in read_very_eager
    self.fill_rawq()
  File "/usr/lib/python2.7/telnetlib.py", line 576, in fill_rawq
    buf = self.sock.recv(50)
socket.error: [Errno 104] Connection reset by peer

```

**Second attempt** - failed within a few seconds

```
/etc/ansible # python telnet_test.py 
25-Feb-2018 06:50:52 read_channel:
25-Feb-2018 06:50:52 write_channel: 

25-Feb-2018 06:50:54 read_channel: 
User Access Verification

Username: 
Username: 
25-Feb-2018 06:50:54 write_channel: mwadmin

25-Feb-2018 06:50:58 read_channel: 
25-Feb-2018 06:50:58 write_channel: 

25-Feb-2018 06:51:00 read_channel: mwadmin
Password: 
25-Feb-2018 06:51:00 write_channel: jump3r5t@rt

25-Feb-2018 06:51:02 read_channel: 
25-Feb-2018 06:51:02 write_channel: 

25-Feb-2018 06:51:04 read_channel: 
Rejected


Username: jump3r5t@rt
Password: 
25-Feb-2018 06:51:04 write_channel: mwadmin

25-Feb-2018 06:51:08 read_channel: 
Rejected


25-Feb-2018 06:51:08 write_channel: 

Traceback (most recent call last):
  File "telnet_test.py", line 22, in <module>
    session_timeout=90)
  File "/usr/lib/python2.7/site-packages/netmiko/ssh_dispatcher.py", line 173, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 160, in __init__
    self.establish_connection()
  File "/usr/lib/python2.7/site-packages/netmiko/base_connection.py", line 640, in establish_connection
    self.telnet_login()
  File "/usr/lib/python2.7/site-packages/netmiko/cisco_base_connection.py", line 128, in telnet_login
    raise NetMikoAuthenticationException(msg)
netmiko.ssh_exception.NetMikoAuthenticationException: Telnet login failed: 10.179.183.1
/etc/ansible # 

```

**Third attempt** - Expected response

```

/etc/ansible # python telnet_test.py 
25-Feb-2018 06:51:20 read_channel: 

User Access Verification

Username: 
25-Feb-2018 06:51:20 write_channel: mwadmin

25-Feb-2018 06:51:24 read_channel: mwadmin
Password: 
25-Feb-2018 06:51:24 write_channel: jump3r5t@rt

25-Feb-2018 06:51:26 read_channel: 
25-Feb-2018 06:51:26 write_channel: 

25-Feb-2018 06:51:28 read_channel: 

kanrou000001>
kanrou000001>
25-Feb-2018 06:51:33 read_channel: 
25-Feb-2018 06:51:41 read_channel: 
25-Feb-2018 06:51:41 write_channel: 

25-Feb-2018 06:51:41 read_channel: 
kanrou000001>
25-Feb-2018 06:51:42 read_channel: 
25-Feb-2018 06:51:50 read_channel: 
25-Feb-2018 06:51:50 read_channel: 
25-Feb-2018 06:51:50 write_channel: 

25-Feb-2018 06:51:50 read_channel: 
25-Feb-2018 06:51:50 read_channel: 
25-Feb-2018 06:51:50 write_channel: 

25-Feb-2018 06:51:51 read_channel: 
kanrou000001>
25-Feb-2018 06:51:51 read_channel: 
25-Feb-2018 06:51:52 read_channel: 
kanrou000001>
25-Feb-2018 06:51:52 In disable_paging
25-Feb-2018 06:51:52 Command: terminal length 0

25-Feb-2018 06:51:52 write_channel: terminal length 0

25-Feb-2018 06:51:52 Pattern is: kanrou000001
25-Feb-2018 06:51:52 read_channel: 
25-Feb-2018 06:51:52 read_channel: 
25-Feb-2018 06:51:52 read_channel: terminal le
25-Feb-2018 06:51:53 read_channel: 
25-Feb-2018 06:51:53 read_channel: 
25-Feb-2018 06:51:54 read_channel: ngth 0
kanrou000001>
25-Feb-2018 06:51:54 Pattern found: kanrou000001 terminal length 0
kanrou000001>
25-Feb-2018 06:51:54 terminal length 0
kanrou000001>
25-Feb-2018 06:51:54 Exiting disable_paging
25-Feb-2018 06:51:54 write_channel: terminal width 511

25-Feb-2018 06:51:54 Pattern is: kanrou000001
25-Feb-2018 06:51:54 read_channel: 
25-Feb-2018 06:51:54 read_channel: 
25-Feb-2018 06:51:54 read_channel: terminal width 511
25-Feb-2018 06:51:55 read_channel: 
25-Feb-2018 06:51:55 read_channel: 
25-Feb-2018 06:51:56 read_channel: 
kanrou000001>
25-Feb-2018 06:51:56 Pattern found: kanrou000001 terminal width 511
kanrou000001>
25-Feb-2018 06:51:57 read_channel: 
25-Feb-2018 06:51:57 write_channel: 

25-Feb-2018 06:51:57 Pattern is: kanrou000001
25-Feb-2018 06:51:57 read_channel: 
25-Feb-2018 06:51:57 read_channel: 
25-Feb-2018 06:51:58 read_channel: 
kanrou000001>
25-Feb-2018 06:51:58 Pattern found: kanrou000001 
kanrou000001>
25-Feb-2018 06:51:58 write_channel: enable

25-Feb-2018 06:51:58 Pattern is: (kanrou000001|ssword)
25-Feb-2018 06:51:58 read_channel: 
25-Feb-2018 06:51:58 read_channel: 
25-Feb-2018 06:51:58 read_channel: enable
Password: 
25-Feb-2018 06:51:58 Pattern found: (kanrou000001|ssword) enable
Password: 
25-Feb-2018 06:51:58 write_channel: invalid

25-Feb-2018 06:51:58 Pattern is: kanrou000001
25-Feb-2018 06:51:58 read_channel: 
25-Feb-2018 06:51:59 read_channel: 
25-Feb-2018 06:51:59 read_channel: 
% Access denied

kanrou000001>
25-Feb-2018 06:51:59 Pattern found: kanrou000001 
% Access denied

kanrou000001>
25-Feb-2018 06:51:59 write_channel: 

25-Feb-2018 06:51:59 Pattern is: kanrou000001
25-Feb-2018 06:51:59 read_channel: 
25-Feb-2018 06:52:00 read_channel: 
25-Feb-2018 06:52:00 read_channel: 
kanrou000001>
25-Feb-2018 06:52:00 Pattern found: kanrou000001 
kanrou000001>
25-Feb-2018 06:52:00 write_channel: 

25-Feb-2018 06:52:00 Pattern is: kanrou000001
25-Feb-2018 06:52:00 read_channel: 
25-Feb-2018 06:52:00 read_channel: 
25-Feb-2018 06:52:01 read_channel: 
kanrou000001>
25-Feb-2018 06:52:01 Pattern found: kanrou000001 
kanrou000001>
25-Feb-2018 06:52:01 exit_config_mode: 
25-Feb-2018 06:52:01 write_channel: exit

Traceback (most recent call last):
  File "telnet_test.py", line 30, in <module>
    raise NetMikoAuthenticationException('{}'.format(err))
netmiko.ssh_exception.NetMikoAuthenticationException: Failed to enter enable mode. Please ensure you pass the 'secret' argument to ConnectHandler.
/etc/ansible # 
```
